src/walker.ts

```diff
@@ -0,0 +1,126 @@
+import {
+  PotentialStep,
+  Pipeline,
+  Conditional,
+  Command,
+  CommandStep,
+  Step,
+} from '.';
+import { getAndCacheDependency } from './conditional';
+export interface Mutators {
+  pipelineFn?: (pipeline: Pipeline) => Promise<Pipeline>;
+  stepFn?: (step: Step) => Promise<Step>;
+  commandFn?: (command: Command) => Promise<Command>;
+}
+
+/**
+ * @param p The pipeline to evaluate
+ * @returns An evaluated pipeline with all conditionals generated if they were accepted
+ */
+export async function evaluatePipeline(pipeline: Pipeline): Promise<Pipeline> {
+  const conditionalCache = new Map<any, any>();
+  const newSteps = [];
+  for (const step of pipeline.steps) {
+    const newStep = await evaluateStep(step, conditionalCache);
+    if (newStep) {
+      newSteps.push(newStep);
+    }
+  }
+  pipeline.steps = newSteps;
+  return pipeline;
+}
+
+async function evaluateStep(
+  step: PotentialStep,
+  conditionalCache: Map<any, any>,
+): Promise<Step | null> {
+  if (step instanceof Conditional) {
+    if (await step.accept()) {
+      step = await getAndCacheDependency(conditionalCache, step);
+    } else {
+      return null;
+    }
+  }
+  const deps = step.dependencies;
+  const newDeps: Set<Step> = new Set();
+  for (const dep of deps) {
+    const newDep = await evaluateStep(dep, conditionalCache);
```

Comment:
This isn't correct: dependencies of steps are forced to be accepted.

Reply:
I just read the commment, and yeah you're right! Looking at the code I don't quite understand this though. 

https://github.com/joscha/buildkite-graph/blob/edb19136627007a927dad591c905db026c0d74a0/src/base.ts#L111
is passing dependencies and effectDependencies to be unwrapped, and

https://github.com/joscha/buildkite-graph/blob/edb19136627007a927dad591c905db026c0d74a0/src/unwrapSteps.ts#L16

only unwraps them if the acceptor is true. Am I misunderstanding this or is there a mismatch between the docs and the code?